### DISCUSSION TITLE
Quality of training results?

### DISCUSSION THREAD

**Discussion by matthiasgeihs** [ggml-org/llama.cpp#1851] [General]
I've been experimenting with `training-text-from-scratch`.
Compared to training on the same data with [nanoGPT](https://github.com/karpathy/nanoGPT), the results seem considerably worse.
(Given roughly the same training time on CPU.)

I've been using the default parameters, which differ between the different code bases (llama.cpp vs nanoGPT).
I've noticed that the loading of the training data is somewhat different as well.

What are your experiences with `llama.cpp` training?

**Comment by xaedes** [comment]
Which nanoGPT configuration and which data preparation did you use?

**Reply by matthiasgeihs** [reply]
Running `nanoGPT` with default CPU parameters for `shakespeare-char` from the README, but using GPT2 encoding.
And then running `training-text-from-scratch` as explained in the example, but with replacing the input text file content (which seems HTML based) with the text file content from nanoGPT shakespeare.

This is only a qualitative feeling that I got. Obviously the default neural net and training parameters are also different.


Model Name | Layers | Heads | Embedding Dimension | Context
-- | -- | -- | -- | --
nanoGPT | 4 | 4 | 128 | 64
train-text-from-scratch | 16 | 8 | 256 | 64

I might run some more experiments with different parameters and report here.

**Reply by matthiasgeihs** [reply]
@xaedes

Did some further experiments...

Dataset: `shakespeare` from nanoGPT
Model/training parameters: 4 heads, 4 layers, dim 128, context 64, batch 12.
Training for roughly 4:30min each.

This is the output I am getting with `training-text-from-scratch`:
```
  no in theords
KUS.:
CICWse
The my face?

My myins-: ITICI in,
And it;:
A a way
H you the III.
The.

I have there.

And and shall you me,
I shall, theying:
DU VFor.

P for that be.
P me.

A be have?
T theorICK?
```

This is the output I am getting with `nanoGPT`:
```
The Lady BOLINGBROKE:
See how he's more both are none:
For what they since, when you must move me,
And bring it to the people, in the cause,
If you have made your friends.

TRANIO:
I will be, good I do:
Might with you, by him: I'll answer you,
Or I, when he shall stay on the night
And then the senate-morrow of the
```

Is there something wrong with my llama.cpp setup?

**Reply by xaedes** [reply]
Same run-time is not a good comparision, because runtime per iteration differs between the implementations.
Instead the number of optimization iterations or the number of seen batches should be equal or similar.

> Is there something wrong with my llama.cpp setup?

Probably not, just need to let it train longer.


I tried to compare two similar configurations with similar number of optimization iterations:

---

### llama.cpp 

The training code I used here is with some new unmerged changes (coming soon) to support gradient clipping, minimum learning rate, applying weight decay to parameters as used in nanoGPT and memory improvements resulting in (probably; didn't compare here) better runtime.

```sh
train-text-from-scratch --vocab-model vocab.bin \
  --seed 1 --ctx 64 --embd 128 --head 4 --layer 4 \
  --checkpoint-in nothing \
  --checkpoint-out chk-checkpointing-shakespeare-128x4.bin \
  --model-out ggml-checkpointing-shakespeare-128x4-f32.bin \ 
  --train-data "shakespeare_linux_nl.txt" \
  -t 8 -b 12 -n 500 --adam_decay 0.1  --adam-iter 1 \
  --print-info-interval 10 --print-details-interval 0 --predict 16
```

Memory usage: ~570mb
Runtime: 9min

Setting `--adam-iter 1` will result in quite some runtime overhead (due to restart of optimizer), but the training is closer to the one in nanoGPT.
With setting it to one, each batch is optimized for only one iteration.

Runtime when setting `-n 250 --adam-iter 2` is 6min 15s, resulting in 500 opt iterations, but only half of the batches -> the output is a bit worse.

Runtime with `-n 32 --adam-iter 16` is 4min 08s, seeing only 32 batches in a total of 512 optimization iterations.

Looks like I really need to implement a way for switching batches and learning schedule directly between iterations, so we don't have to rely on this `--adam-iter` parameter...

Output of `main -s 1 -c 64 -m ggml-checkpointing-shakespeare-128x4-f32.bin`
<details>

```txt


  it that theest,
ThETKING
BENTIO:
O:
NCEy, theelf you his man, and your house.

And so;
Thar.

I am the soul, a lord
With be me? I thou a wan thou be.

HCE:
WhKEll him's not.

My lORMERCoth the more, that you is thou be be be my duLO
And for he her in my own!

NENRYLOINO:
The name, and a brother, 'ET:
The brother.

What it of ourest his father, in will it is the house,
H:
And the true to the duLO: to, and he's so of your lord,
To that him, to you,
ThARD:
FMEYSTENARINGGLOO:
NENA:
With my king, my my man?

IENENCEing.

LOLET:
PER:
BENTUS:
I are I have her of thou your lord:
And the bLOIING RICH:
That that this,
L VI:
To have his lord;
My king's not a wou.

And I me so,
And be the lord's IUS:
I say, to my more not he thee is you is it,
For my lord of her,
To his be it.

H: my king, she a and that the name,
```
</details>


Output of `main -s 1 -c 64 -m ggml-checkpointing-shakespeare-niter2-128x4-f32.bin`
<details>

```txt
  thy blood, by the soul.
ThRET:
I'll my heart his son of so, and we had all the lord;
WhICO:
What
And thou- life to be a way, which thed, the not to her and it are you's- mother
He,
PETYENIO
K:
I'd, for not my mother.

And to the wife, I is your man?

The lIC:
Nou good heart of the death!

A:
And to be so, but I am you this to you, I will not your life and your good life.

And our thing with her, if the life.


And they you'd.
What:
A:
GUS:
And have the my life, you's,
The lord, his he that him
WARA:
But as we't,
If you's not that to have a the father's in now.

HORHARD:
To the thing and be that you this the both a name, sir.

And I am I the blood'd,
Thes.

IUS:
And he are you's;
MENIO:
And the lord me in thy lord, but him,
And thou the father!

Which my good?

The way:
```
</details>

Output of `main -s 1 -c 64 -m ggml-checkpointing-shakespeare-niter16-128x4-f32.bin`
<details>

```txt
  good thaneth be it be theoundsly?
G aIONE I's for and crown.

Y:
The;
To he'd with, thatIO:
TUCy.

I with, a aodTAGoul the old lord me and;
WeUS:
LUCULI! and with in his body
Whats and the friends
The thanms it this,
And youred, not?
N will is,
That?
FAN:
To's of;
G aREI is.

Now, and a done:
I is to,;
But her myay!

SHARDAnd?
MIO:
I'd have sir, and for;
BR his body

KINGThION you the very; and life.

LUCIO:
That for inin!

O:
N, in is.

The grace;
I be?
My lord, my mind you is youray's to all,
The word you' the life:
SINGL last,
Y:
MIO:
And, with me:
LUCIO:
The friendsUS:
DUCHO:
GEN:
That?
B;
I'dSTUS:
The th's.

I'll him this theirish the life it,,
C and for is with, you thee it aKEINO:
LUCy
Why and's of the life too thousandms it is,
```
</details>

To get some better output I let it train longer with 512 batches and 8 iterations per batch (`-n 512 --adam-iter 8`, runtime ~36min):

Output of `main -s 1 -c 64 -m ggml-checkpointing-shakespeare-n512-niter8-128x4-f32.bin`
<details>

```txt
  not we have as I
The foe that the Duke of my soul to a
He, which his son: they should will be
For thou have more than a father; and the besty
That is the good father: I would not; then
Which you the people and be thyself;
And let me have that the Duke of this name,
His. But you would you with you,
I pray him, thou, we should be
To his son.

BENCESTER:
O, sir, my lord, and not, but I am
So he is the brother's my father.

HAMILLO:
HAMILLO:
The world is the duke hath done, in a
We say my master, so long to this,
Ays his father!

KING RICHath.

KATHARWARD II:
BAPUCESTAN:
IOLUMNIA:
What, thou art not our mother of the time's,
Would I is my brother of the blood; to be
He shall will I have not a king, but my grace!

GLOULIET:
Away of the people:
Thou be a wilt to do.

KATHARINCENTIO:
I'll be a gods! for him, sir, or I say,
Sir-gain's the good; and
Which, my wife, no king, sir?

RICINCE:
And, and I do you shall not, he wasd with your crown.

ROMEO:
And if thou have to thine so he are in it:
```
</details>

---

### nanoGPT

Saving `data\shakespeare\input.txt` with linux line endings and then prepare the data:

```python
python data\shakespeare\prepare.py
```

In `config\train_shakespeare.py`:

```python
[...]
dataset = 'shakespeare'
gradient_accumulation_steps = 1
batch_size = 12
block_size = 64
n_layer = 4
n_head = 4
n_embd = 128
dropout = 0.0
learning_rate = 1e-3
max_iters = 500
lr_decay_iters = 1000
min_lr = 1e-4
beta1 = 0.9
beta2 = 0.999
weight_decay = 0.1
dtype = 'float32'
warmup_iters = 100
device = 'cpu'
compile = False
```

Training:
```sh
python train.py config\train_shakespeare.py
```

Memory usage: up to ~780mb
Runtime: 4min including 46s for computing validation loss. So 3:14 for training itself.

Output of `python sample.py --device=cpu --out_dir=out-shakespeare`

<details>

```txt
Let you did in our request;

Is my hour to this must
Is me no to know the gracious lord;
For now!
MENENIUS:
To find my lord; you would
IET:
Well, and now's
That I'll tell me the more to a traitor at his Duke of,
And nothing so!


What shall will been, and the house of the soul,
For I have what Edward, and, how doth not:
When they for honour to do.

FirstUMABELLA:
As,
Have he hath I have a king's to a king'st Ist we shall make you,
With the truth or have:

to, but on these the life, my hand,
Whall I would to be to our state is a sea of this,

And take the time and one that would have a world?


SecondOMUMNurse:
You'll would look you.

Nurse:
QUEENI am we have be it is not have not they will hear
Here hath the part a good a
That that I have a king? a tender time's York: go?




SareASTOUCALUS:
IUS:
Have, he should besam, my lord, I, sir,
In this arm of love:
From me.

DUCHESS OF YORK:

What is't:
Away upon Rome.


ISABETH:
I'll had, I have so's banvoling thy matter:
I must I are be thy people,
And I am no.
Well, marry you that thou havestst in thee from the king,
To make one in our man I have to
AICK:
You had
Have not too more.

Third Gentleman:
For how he will not no more place, nor I--there's dead is Anaheim in our king?


DUKE OF YORK:
CALUS:
To are for sweet father is so our king:
'Tis, sir, are you have these Richard,
And for my sword
And then thoth he had a own.


FirstARET:
For he duke is a people
for't with him of me, good
```

</details>

---

Output from nanoGPT is still better, but the difference is not as big as in your results. 
Also note that we compare two different model architectures with two different tokenizations: n_vocab=32001 for llama.cpp and n_vocab=50257 for GPT2(nanoGPT).

A bit closer comparision would be with both tokenizations on byte/character level (256 tokens, one for each extended-ascii byte value).
Would be nice to have such a tokenizer available for use in llama.cpp, but I couldn't find one in the webs.

Llama runtime performance can probably still be improved - I especially hope for tiled matrix multiplication (https://github.com/users/ggerganov/projects/7/views/1?pane=issue&itemId=31669483) to improve it some more.

**Reply by matthiasgeihs** [reply]
Thanks a lot for investigating this. I'll have a closer look at your results and maybe try to reproduce when I find the time. 🙏

**Reply by Green-Sky** [reply]
[ggml-bytewise-vocab.bin.zip](https://github.com/ggerganov/llama.cpp/files/11918214/ggml-bytewise-vocab.bin.zip)
generated using https://gist.github.com/Green-Sky/aae372314f981d14b00852c34985c93a

**Reply by Green-Sky** [reply]
i hope it works, atleast it loads

**Reply by xaedes** [reply]
Yes, it works :)

**Reply by matthiasgeihs** [reply]
Maybe 1 additional note / question @xaedes: I was kind of assuming that `train-text-from-scratch` efficiency would beat `nanoGPT`/`pytorch` training efficiency on `cpu` (reasoning: basically because llama.cpp inference is so convincingly fast). hence i thought at same runtime, `train-text-from-scratch` results should be at least comparable, if not better...

According to your recent comment, this assumption seems to be wrong, meaning `train-text-from-scratch` is currently not more efficient than conventional `pytorch` training on `cpu`. is that correct? can you maybe shed more light on how the two different approaches compare?

**Comment by aseok** [comment]
here's output of training-text-from-scratch with latest default parameters from examples page, seems it's exact chunk of input text:
Generating 16 tokens.
 in hue,
Could make me any summer's story tell:
Or from their proud lap pluck them where they grew:  
Nor did I wonder at the lily's white,
Nor praise the deep vermilion in---
 very veryist in my love,
Which to be
Thou

**Comment by the-crypt-keeper** [comment]
1 training epoch output:

```
Thy still bladoth middle self I
For thou againing their mort-read.
There not my seek not?
M kind,
The beauty in his days in the soul is so.
And-atiencealill rose


And my love, and notich shalls thoughts I to this he love,
Why his rose,
O, and I to his strong thou shall beom,


Wh are org defence the maj sacredetses me since a kind,

Since.
But love,
And having,

But that thouet night.
```

perplexity on input text:
```
[1]322.9544,[2]344.1555,[3]384.6766,[4]408.3293,[5]445.5106,[6]410.7461,[7]415.0346,[8]393.5498,[9]405.0035,[10]391.7012,[11]396.9100,[12]404.1142,[13]386.9103,[14]386.8066,[15]380.7482,[16]382.7702,[17]384.9585,[18]371.8133,[19]381.9014,[20]373.3629,[21]380.8767,[22]381.7560,[23]398.5143,[24]402.8592,[25]398.6330,[26]398.5414,[27]393.8434,[28]391.7720,[29]389.1273,[30]387.1373,[31]383.4092,[32]379.2208,[33]377.6578,[34]382.7226,[35]384.2917,[36]382.4465,[37]376.8589,[38]373.1323,[39]373.4557,[40]375.5122,[41]380.7512,[42]379.9440,[43]379.6943,[44]380.9680,[45]380.9655,[46]378.0578,[47]376.2581,[48]375.4618,[49]371.8680,[50]373.0968,[51]378.1924,[52]374.2484,[53]377.5955,
```

2nd iteration of training output:

```
 .

And I doth fairest it that I I'st,
Mine of all thy self-air, and to that I in the world to be be are's mays sight to mine then of love,

Whoseillalt his golden am I do,
For,
Inou wy beauty of his salan of your praise,
Theup love.
But then not as eyes as my self the verse,
Since thouoe, by ais with-ire,

Who best to my self is my love'sid are in hispass thy beauty's
```

2nd iteration perplexity:

```
[1]160.8362,[2]167.8282,[3]193.6216,[4]217.0640,[5]246.7007,[6]222.9784,[7]238.6129,[8]219.8489,[9]224.8858,[10]215.0258,[11]218.3044,[12]230.5945,[13]219.4109,[14]219.0968,[15]214.7964,[16]217.4400,[17]215.9933,[18]210.9362,[19]220.2809,[20]213.3470,[21]220.1183,[22]220.8794,[23]233.6096,[24]238.2936,[25]236.2420,[26]237.0931,[27]232.7523,[28]231.1317,[29]232.6202,[30]231.0412,[31]229.6245,[32]227.1756,[33]224.2280,[34]230.1064,[35]231.7224,[36]229.5500,[37]225.8269,[38]222.3605,[39]222.1848,[40]224.6341,[41]230.0675,[42]227.1575,[43]228.5041,[44]228.3183,[45]229.0491,[46]227.9974,[47]227.8194,[48]227.1214,[49]224.7755,[50]224.6679,[51]229.1267,[52]225.9887,[53]229.7836,
```

3rd iteration output:

```
  of his society to looks this our his rose is,
And you on the dro wide, since his middle age,
Incertitled in your self by day'seds,
For blessed than my hath thee,
Therefore are but what blessed than my truth suppressed:
Thy black-day, and so are past the rhancyit,
For blessed than mythough she so solemn
And for I in thy love, he not for that which he doth say,
The it self by night, he will not every bluns,
Therefore, and so both to bl
```

3rd iteration perplexity:
```
[1]330.3301,[2]335.3353,[3]381.7894,[4]450.8096,[5]509.5158,[6]436.1262,[7]485.2961,[8]454.9817,[9]454.4107,[10]427.2342,[11]446.5920,[12]480.8394,[13]454.9135,[14]462.6143,[15]457.2441,[16]458.1195,[17]458.5167,[18]422.7817,[19]446.6622,[20]425.9623,[21]441.9038,[22]444.7060,[23]472.0163,[24]481.6250,[25]478.6315,[26]486.0071,[27]472.7340,[28]469.3013,[29]477.4521,[30]469.9810,[31]466.6954,[32]461.6271,[33]451.5267,[34]464.8970,[35]469.3584,[36]460.3976,[37]446.8347,[38]439.6502,[39]441.1156,[40]441.3072,[41]450.1176,[42]440.6842,[43]439.7314,[44]439.9365,[45]438.7138,[46]430.9184,[47]431.8198,[48]432.9292,[49]427.4609,[50]428.2032,[51]437.8436,[52]432.3350,[53]439.6829,
```

**Comment by daboe01** [comment]
i have trained 100MB medical records from scratch with the aim of text competition. training was 1 week on a macpro (48GB ram).
unfortunately, no useful text is generated from this model (although the generated tokens after training looked promising).

the same training data give excellent results upon LORA finetuning a 7b llama (pytorch, GPU cluster).
